src/TopSurface.h

```diff
@@ -0,0 +1,59 @@
+//Copyright (c) 2017 Ultimaker B.V.
+//CuraEngine is released under the terms of the AGPLv3 or higher.
+
+#ifndef TOPSURFACE_H
+#define TOPSURFACE_H
+
+#include "utils/polygon.h" //For the polygon areas.
+#include "sliceDataStorage.h" //For the input mesh.
+
+namespace cura
+{
+
+class SliceMeshStorage;
+
+class TopSurface
```

Comment:
This architecture is rather different from how features are implemented elsewhere. Generally the polygon generation phase is separated from the path generation phase. This is because classes in the engine are generally wrappers for computation.

I've made the same mistake when implementing the prime tower. There I used the same class for the polygon generation phase as the path planning phase as well. The problem was that class members can then only be used for either phase. The prime tower class has private members used in the polygon generation phase, but all parameters of the computation of the path generation phase had to be passed via arguments.

I think your architecture might be better than what we had, but it does introduce a new type of architecture in the engine, which is already chaotic in its architecture. I think we should discuss the architecture and decide together whether the architecture of this class is the type of architecture we want all print features to have eventually or whether this print feature should conform to existing architecture.